src/runmode-dpdk.c

```diff
@@ -0,0 +1,1295 @@
+    if (retval != 0) {
+        SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) during configuring the device (port %u)",
+                retval, iconf->port_id);
+        SCReturnInt(retval);
+    }
+
+    retval = rte_eth_dev_adjust_nb_rx_tx_desc(
+            iconf->port_id, &iconf->nb_rx_desc, &iconf->nb_tx_desc);
+    if (retval != 0) {
+        SCLogError(SC_ERR_DPDK_INIT,
+                "Error (err=%d) during adjustment of device queues descriptors (port %u)", retval,
+                iconf->port_id);
+        SCReturnInt(retval);
+    }
+
+    retval = iconf->flags & DPDK_MULTICAST ? rte_eth_allmulticast_enable(iconf->port_id)
+                                           : rte_eth_allmulticast_disable(iconf->port_id);
+    if (retval == -ENOTSUP) {
+        retval = rte_eth_allmulticast_get(iconf->port_id);
+        // when multicast is enabled but set to disable or vice versa
+        if ((retval == 1 && !(iconf->flags & DPDK_MULTICAST)) ||
+                (retval == 0 && (iconf->flags & DPDK_MULTICAST))) {
+            SCLogError(SC_ERR_DPDK_CONF,
+                    "Allmulticast setting of port (%" PRIu16
+                    ") can not be configured. Set it to %s",
+                    iconf->port_id, retval == 1 ? "true" : "false");
+        } else if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) Unable to get multicast mode on port %u",
+                    retval, iconf->port_id);
+            SCReturnInt(retval);
+        }
+
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) Unable to get multicast mode on port %u",
+                    retval, iconf->port_id);
+            SCReturnInt(retval);
+        }
+    } else if (retval < 0) {
+        SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) when en/disabling multicast on port %u",
+                retval, iconf->port_id);
+        SCReturnInt(retval);
+    }
+
+    retval = iconf->flags & DPDK_PROMISC ? rte_eth_promiscuous_enable(iconf->port_id)
+                                         : rte_eth_promiscuous_disable(iconf->port_id);
+    if (retval == -ENOTSUP) {
+        retval = rte_eth_promiscuous_get(iconf->port_id);
+        if ((retval == 1 && !(iconf->flags & DPDK_PROMISC)) ||
+                (retval == 0 && (iconf->flags & DPDK_PROMISC))) {
+            SCLogError(SC_ERR_DPDK_CONF,
+                    "Promiscuous setting of port (%" PRIu16 ") can not be configured. Set it to %s",
+                    iconf->port_id, retval == 1 ? "true" : "false");
+        } else if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) Unable to get promiscuous mode on port %u",
+                    retval, iconf->port_id);
+            SCReturnInt(retval);
+        }
+    } else if (retval < 0) {
+        SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) when enabling promiscuous mode on port %u",
+                retval, iconf->port_id);
+        iconf->DerefFunc(iconf);
```

Comment:
Should `DerefFunc` be used on all error exits?

Reply:
Thanks for the comment, I removed the `DerefFunc` because, in case of an error in this function, the caller (`ParseDpdkConfigAndConfigureDevice`) dereferences the config.

Reply:
I considered `FatalError` approach even in the device configuration phase but as I am interacting with the device (in the configuration phase) I wanted to execute rte_eal_cleanup in case of an error. I placed the rte_eal_cleanup in the function `ParseDpdkConfigAndConfigureDevice` where I also dereference/free the allocated configuration structure. It is not necessary but I thought it is a good practice to clean up even on error exits, especially if it is easily achievable.